src/zmpl/Template.zig

```diff
@@ -409,33 +500,94 @@ fn compileMarkupLine(self: *Self, line: []const u8) ![]const u8 {
     return markup_line.compile();
 }
 
-fn compileRawLine(self: *Self, line: []const u8, chomp: bool) ![]const u8 {
-    const chomped = if (chomp) chompString(line) else line;
+fn compileRaw(self: *Self, string: []const u8, chomp: bool) ![]const u8 {
+    const chomped = if (chomp) chompString(string) else string;
+    const escaped_backslash = try std.mem.replaceOwned(
+        u8,
+        self.allocator,
+        chomped,
+        "\"",
+        "\\\"",
+    );
+    defer self.allocator.free(escaped_backslash);
+
+    const escaped_linebreak_rn = try std.mem.replaceOwned(
+        u8,
+        self.allocator,
+        escaped_backslash,
+        "\r\n",
+        "\\r\\n",
+    );
+    defer self.allocator.free(escaped_linebreak_rn);
+
+    const escaped_linebreak_n = try std.mem.replaceOwned(
+        u8,
+        self.allocator,
+        escaped_linebreak_rn,
+        "\n",
+        "\\n",
+    );
+    defer self.allocator.free(escaped_linebreak_n);
+
+    const compiled = escaped_linebreak_n;
+
     return std.mem.join(
         self.allocator,
         "",
         &[_][]const u8{
             "try zmpl.write(\"",
-            try std.mem.replaceOwned(u8, self.allocator, chomped, "\"", "\\\""),
+            compiled,
             "\\n\");",
         },
     );
 }
 
 fn compileZigLine(self: *Self, line: []const u8) ![]const u8 {
-    _ = self;
-    return line;
+    return try self.allocator.dupe(u8, line);
 }
 
 fn escapeText(self: *Self, text: []const u8) ![]const u8 {
     _ = self;
     return text;
 }
 
+fn escapeChar(char: u8) []const u8 {
+    if (char == '\r') return "\\r";
+    if (char == '\n') return "\\n";
+    if (char == '"') return "\"";
+    if (char == '\\') return "\\\\";
+
+    return &[_]u8{char};
+}
+
 fn chompString(string: []const u8) []const u8 {
     if (std.mem.endsWith(u8, string, "\r\n")) {
         return string[0 .. string.len - 3];
     } else if (std.mem.endsWith(u8, string, "\n")) {
         return string[0 .. string.len - 2];
     } else return string;
 }
+
+fn isWhitespace(string: []const u8) bool {
+    for (string) |char| {
+        if (char == ' ' or char == '\t') continue;
+        return false;
+    }
+    return true;
+}
+
+fn clearDanglingWhitespace(buf: *std.ArrayList(u8)) void {
```

Comment:
Is it important that only whitespace prepended by a newline is cleared? Otherwise, the implementation could be shortened to something like 

```zig
    const whitespace = [_]u8{ ' ', '\t', '\r', '\n'};
    var end = buf.items.len;
    while (end > 0 and std.mem.indexOfScalar(u8, &whitespace, buf.items[end - 1]) != null) : (end -= 1) {
        _ = buf.pop();
    }
```

Whether that's a better version is of course subject to personal preference.

If really only whitespace after a newline should be removed, there doesn't seem to be a test that covers the "wasn't removed" case.

Reply:
Yeah, the use case here is:
```html
<#>hello</#>

   <#>
   hello
   </#><span>yooo</span>
````
I did wonder if I'm imagining scenarios a little too much but I think it's reasonable to expect to be able to add stuff after the tag, and I thought it would be highly surprising if we magically removed stuff. We could raise an error, but then you couldn't add stuff directly after a raw tag.

FWIW I was extremely tired and had a horrible headache when I wrote this.

Either way, I think your point that the test coverage isn't there is the most important thing - I'll make that a TODO.

Reply:
```
zig-cache/o/5a5bab81bc996a3adb36c081d1f0a0e9/templates/example_with_multi_line_fragment_tag_and_trailing_content.zmpl:11:6: error: expected ';' after statement
some trailing content
```
lol